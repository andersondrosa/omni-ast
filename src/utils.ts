import { ObjectExpressionToJSON } from "./JsonGenerate";
import { Node, Property } from "../src/types";
import { lit, objectExpression } from "./builders";
import { serialize } from "./generators";

type Options = { pre?; post?; skipProperty? };

export function simpleTraverse(root, options: Options = {}) {
  var pre = options.pre;
  var post = options.post;
  var skipProperty = options.skipProperty;

  function visit(node, parent, prop?, idx?) {
    if (!node || typeof node.type !== "string") return;

    var res = undefined;
    if (pre) res = pre(node, parent, prop, idx);

    if (res !== false) {
      for (const prop in node) {
        if (skipProperty ? skipProperty(prop, node) : prop[0] === "$") continue;
        const child = node[prop];
        if (Array.isArray(child)) {
          for (var i = 0; i < child.length; i++) visit(child[i], node, prop, i);
        } else visit(child, node, prop);
      }
    }

    if (post) post(node, parent, prop, idx);
  }

  visit(root, null);
}

export const Mock = new Proxy({ fn: {} } as any, {
  get(target, name, receiver) {
    return { type: "Identifier", name };
  },
});

const isObject = (item) =>
  item && typeof item === "object" && !Array.isArray(item);

export function find(node, match) {
  //
  if (typeof node != "object" || node === null) return;

  const recursive = (object: Object) => {
    //
    if (match(object)) return object;

    if (Array.isArray(object)) {
      for (const val of object) {
        const res = recursive(val);
        if (res) return res;
      }
      return;
    }

    for (const i in object) {
      const value = object[i];
      if (typeof value != "object" || value === null) continue;
      const res = recursive(value);
      if (res) return res;
    }
  };

  return recursive(node);
}

export function mutate(object, match, modifier) {
  const findAndModify = (object) => {
    if (isObject(object)) {
      console.log(typeof match);
      if (match(object)) return modifier(object);
      const newObj = {};
      let changed = false;
      for (const [k, v] of Object.entries(object)) {
        const newValue = findAndModify(v);
        newObj[k] = newValue;
        if (newValue !== v) changed = true;
      }
      return changed ? newObj : object;
    }
    if (Array.isArray(object)) {
      const newArray = object.map(findAndModify);
      return newArray.some((item, index) => item !== object[index])
        ? newArray
        : object;
    }
    return object;
  };
  return findAndModify(object);
}

const denied = ["start", "end", "generator"];
const hideIfNull = ["async", "id", "label", "computed", "optional"];

export function cleanAST(ast): any {
  if (typeof ast != "object" || ast === null) return ast;
  const res = {};
  for (const key in ast) {
    if (denied.includes(key)) continue;
    const value = ast[key];
    if (hideIfNull.includes(key)) {
      // console.log(">>", key, value);
      if (!value) continue;
    }
    if (typeof value != "object") {
      res[key] = value;
      continue;
    }
    if (Array.isArray(value)) {
      res[key] = value.map((x) => cleanAST(x));
      continue;
    }
    res[key] = cleanAST(value);
  }
  return res;
}

export function parseOmniAST(ast: Node) {
  //
  if (typeof ast != "object" || ast === null) return ast;

  if (Array.isArray(ast)) return ast.map(parseOmniAST);

  if (!ast.hasOwnProperty("type")) return ast;

  if (typeof ast["type"] != "string")
    throw Error("Invalid AST. Type is invalid");

  if (ast.type == "ObjectExpression")
    return { type: "JsonExpression", body: ObjectExpressionToJSON(ast) };

  const omniAst = {};
  for (const key in ast) {
    omniAst[key] = parseOmniAST(ast[key]);
  }
  return omniAst;
}

export function parseAST(value: Node) {
  //
  const astFindJSON = (ast: Node) => {
    //
    if (Array.isArray(ast)) return ast.map(astFindJSON);

    if (ast.type == "JsonExpression") return jsonToAST(ast.body);

    const _ast: Node = {} as Node;
    for (const key in ast) {
      _ast[key] =
        typeof ast[key] === "object" && ast[key] != null
          ? astFindJSON(ast[key])
          : ast[key];
    }
    return _ast;
  };

  const jsonToAST = (value) => {
    //
    if (typeof value != "object") return lit(value);

    if (Array.isArray(value))
      return { type: "ArrayExpression", elements: value.map(jsonToAST) };

    if (value.type == "#AST") return astFindJSON(value.body);

    const properties: Property[] = [];

    for (const key in value) {
      const row = value[key];
      properties.push({
        type: "Property",
        key: lit(key),
        value:
          typeof row != "object" || row === null ? lit(row) : jsonToAST(row),
        computed: false,
        kind: "init",
        method: false,
        shorthand: false,
      });
    }
    return objectExpression(properties);
  };

  return astFindJSON(value);
}

export const tryGenerate = (ast) => {
  try {
    return serialize(ast);
  } catch (e) {
    return `[${e.message}]`;
  }
};
