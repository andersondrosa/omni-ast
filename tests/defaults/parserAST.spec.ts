import { describe, expect, it } from "vitest";
import { tokenizer } from "../utils/tokenizer";
import { acornParse } from "../utils/acornParse";
import {
  clearAST,
  simplify,
  parseAST,
  builders,
  ast,
  json,
  generate,
} from "../../src";

const { identifier } = builders;
const b = builders;

const log = (x) => console.dir(x, { depth: 20 });

describe("Parsers test", () => {
  //
  it("Should test simplified AST x complex AST", () => {
    //
    const script = `({ 
      json: "here", 
      jsonFunction: jsAgain({ json: "again" }) 
    })`;

    const acornAST = acornParse(script).body[0].expression;

    const cleanAcornAST = clearAST(acornAST);

    const simpleAST = simplify(cleanAcornAST);

    expect(simpleAST).toMatchObject(
      json({
        json: "here",
        jsonFunction: ast({
          type: "CallExpression",
          callee: { type: "Identifier", name: "jsAgain" },
          arguments: [{ type: "JsonExpression", body: { json: "again" } }],
        }),
      })
    );

    const AST = parseAST(simpleAST);

    expect(AST).toMatchObject(cleanAcornAST);

    expect(AST).toMatchObject({
      type: "ObjectExpression",
      properties: [
        {
          type: "Property",
          key: identifier("json"),
          value: { type: "Literal", value: "here", raw: '"here"' },
          computed: false,
          kind: "init",
          method: false,
          shorthand: false,
        },
        {
          type: "Property",
          key: identifier("jsonFunction"),
          value: {
            type: "CallExpression",
            callee: identifier("jsAgain"),
            arguments: [
              {
                type: "ObjectExpression",
                properties: [
                  {
                    type: "Property",
                    key: identifier("json"),
                    value: b.literal("again"),
                    computed: false,
                    kind: "init",
                    method: false,
                    shorthand: false,
                  },
                ],
              },
            ],
          },
          computed: false,
          kind: "init",
          method: false,
          shorthand: false,
        },
      ],
    });
  });

  it("Hybrid", () => {
    const script = `({ 
      json: "here", 
      jsonFunction: jsAgain({ json: "again" }) 
    })`;

    const realAST = acornParse(script).body[0].expression;

    expect(realAST).toMatchSnapshot();

    const cleanAcornAST = clearAST(realAST);

    expect(cleanAcornAST).toMatchSnapshot();

    const omniAST = simplify(cleanAcornAST);

    const AST = parseAST(omniAST);

    // expect(AST).toMatchObject(cleanAcornAST);

    const generated = `(${generate(omniAST)})`;

    expect(tokenizer(script)).toMatchObject(tokenizer(generated));
  });

  it("AssignmentExpression", () => {
    //
    const AST = b.assignmentExpression(
      "=",
      b.identifier("script"),
      b.arrowFunctionExpression([], b.literal("ok"))
    );

    const simpleAST = simplify(AST);

    expect(simpleAST).toEqual({
      left: { name: "script", type: "Identifier" },
      operator: "=",
      right: {
        body: { raw: '"ok"', type: "Literal", value: "ok" },
        expression: true,
        params: [],
        type: "ArrowFunctionExpression",
      },
      type: "AssignmentExpression",
    });

    expect(parseAST(simpleAST)).toMatchObject(AST);
  });

  it("FunctionExpression", () => {
    //
    const AST = b.functionExpression(
      b.identifier("name"),
      [],
      b.blockStatement([b.expressionStatement(b.identifier("script"))]),
      true
    );

    const simpleAST = simplify(AST);
    log(simpleAST);

    expect(simpleAST).toEqual({
      type: "FunctionExpression",
      params: [],
      body: {
        type: "BlockStatement",
        body: [
          {
            type: "ExpressionStatement",
            expression: { type: "Identifier", name: "script" },
          },
        ],
      },
      id: { type: "Identifier", name: "name" },
      async: true,
    });

    expect(parseAST(simpleAST)).toMatchObject(AST);
  });

  it("MemberExpression", () => {
    //
    const AST = b.memberExpression(b.identifier("name"), b.literal("value"));

    log(AST);

    const simpleAST = simplify(AST);

    log(simpleAST);
    expect(simpleAST).toEqual({
      type: "MemberExpression",
      object: { type: "Identifier", name: "name" },
      property: { type: "Literal", value: "value", raw: '"value"' },
    });

    const restored = parseAST(simpleAST);
    log(restored);

    expect(restored).toMatchObject(AST);
  });
});
