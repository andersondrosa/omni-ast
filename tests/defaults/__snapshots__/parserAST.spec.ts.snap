// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Parsers test > Hybrid 1`] = `
Node {
  "end": 78,
  "properties": [
    Node {
      "computed": false,
      "end": 22,
      "key": Node {
        "end": 14,
        "name": "json",
        "start": 10,
        "type": "Identifier",
      },
      "kind": "init",
      "method": false,
      "shorthand": false,
      "start": 10,
      "type": "Property",
      "value": Node {
        "end": 22,
        "raw": "\\"here\\"",
        "start": 16,
        "type": "Literal",
        "value": "here",
      },
    },
    Node {
      "computed": false,
      "end": 71,
      "key": Node {
        "end": 43,
        "name": "jsonFunction",
        "start": 31,
        "type": "Identifier",
      },
      "kind": "init",
      "method": false,
      "shorthand": false,
      "start": 31,
      "type": "Property",
      "value": Node {
        "arguments": [
          Node {
            "end": 70,
            "properties": [
              Node {
                "computed": false,
                "end": 68,
                "key": Node {
                  "end": 59,
                  "name": "json",
                  "start": 55,
                  "type": "Identifier",
                },
                "kind": "init",
                "method": false,
                "shorthand": false,
                "start": 55,
                "type": "Property",
                "value": Node {
                  "end": 68,
                  "raw": "\\"again\\"",
                  "start": 61,
                  "type": "Literal",
                  "value": "again",
                },
              },
            ],
            "start": 53,
            "type": "ObjectExpression",
          },
        ],
        "callee": Node {
          "end": 52,
          "name": "jsAgain",
          "start": 45,
          "type": "Identifier",
        },
        "end": 71,
        "optional": false,
        "start": 45,
        "type": "CallExpression",
      },
    },
  ],
  "start": 1,
  "type": "ObjectExpression",
}
`;

exports[`Parsers test > Hybrid 2`] = `
{
  "properties": [
    {
      "computed": false,
      "key": {
        "name": "json",
        "type": "Identifier",
      },
      "kind": "init",
      "shorthand": false,
      "type": "Property",
      "value": {
        "raw": "\\"here\\"",
        "type": "Literal",
        "value": "here",
      },
    },
    {
      "computed": false,
      "key": {
        "name": "jsonFunction",
        "type": "Identifier",
      },
      "kind": "init",
      "shorthand": false,
      "type": "Property",
      "value": {
        "arguments": [
          {
            "properties": [
              {
                "computed": false,
                "key": {
                  "name": "json",
                  "type": "Identifier",
                },
                "kind": "init",
                "shorthand": false,
                "type": "Property",
                "value": {
                  "raw": "\\"again\\"",
                  "type": "Literal",
                  "value": "again",
                },
              },
            ],
            "type": "ObjectExpression",
          },
        ],
        "callee": {
          "name": "jsAgain",
          "type": "Identifier",
        },
        "optional": false,
        "type": "CallExpression",
      },
    },
  ],
  "type": "ObjectExpression",
}
`;
